src/crunch/crn_decomp.rs

```diff
@@ -0,0 +1,224 @@
+use super::CrnFormat;
+use super::crn_consts::*;
+use super::crn_unpacker::*;
+extern crate alloc;
+
+// #[repr(C)]
+// pub struct crn_file_info{
+//     m_struct_size: u32,
+//     m_actual_data_size: u32,
+//     m_header_size: u32,
+//     m_total_palette_size: u32,
+//     m_tables_size: u32,
+//     m_levels: u32,
+//     m_level_compressed_size: [u32; cCRNMaxLevels as usize],
+//     m_color_endpoint_palette_entries: u32,
+//     m_color_selector_palette_entries: u32,
+//     m_alpha_endpoint_palette_entries: u32,
+//     m_alpha_selector_palette_entries: u32
+// }
+
+// impl crn_file_info{
+//     pub fn default() -> crn_file_info{
+//         return crn_file_info { 
+//             m_struct_size: core::mem::size_of::<crn_file_info>() as u32,
+//             m_actual_data_size: 0,
+//             m_header_size: 0,
+//             m_total_palette_size: 0,
+//             m_tables_size: 0,
+//             m_levels: 0,
+//             m_level_compressed_size: [0; cCRNMaxLevels as usize],
+//             m_color_endpoint_palette_entries: 0,
+//             m_color_selector_palette_entries: 0,
+//             m_alpha_endpoint_palette_entries: 0,
+//             m_alpha_selector_palette_entries: 0
+//         }
+//     }
+// }
+
+#[repr(C)]
+pub struct CrnPackedUint<const N: usize>{
+    pub m_buf: [u8; N]
+}
+
+// no-std, so we can not use std::ops
+impl<const N: usize> CrnPackedUint<N>{
+    pub fn assign_from_buffer(&mut self, other: &[u8]){
+        self.m_buf.copy_from_slice(&other[0..N])
+    }
+
+    pub fn cast_to_uint(&mut self) -> u32{
+        return match N {
+            1 => self.m_buf[0] as u32,
+            2 => u16::from_be_bytes([self.m_buf[0], self.m_buf[1]]) as u32,
+            3 => (self.m_buf[0] as u32) << 16 | u16::from_be_bytes([self.m_buf[1], self.m_buf[2]]) as u32,
+            4 => u32::from_be_bytes([self.m_buf[0], self.m_buf[1], self.m_buf[2], self.m_buf[3]]),
+            _ => panic!("Packed integer can hold a 4 byte buffer at max!")
+        }
+    }
+
+}
+
+impl<const N: usize> Default for CrnPackedUint<N>{

```

Comment:
replace with `#[derive(Default)]`

Reply:
Rust's Default trait does not support generic constants, so it can't be replaced with a derive.